fastavro/_write_common.py

```diff
@@ -0,0 +1,18 @@
+
+class ValidationException(Exception):
+    def __init__(self, datum, schema, field):
+        message = '{field} is {datum} of type ' \
+                  '{given_type} expected {schema}'. \
+            format(datum=datum, given_type=type(datum),
+                   schema=schema, field=field)
+        super(ValidationException, self).__init__(message)
+        self.message = message
+        self.datum = datum
+        self.schema = schema
+        self.field = field
+
+
+class ValidationErrors(ValidationException):
```

Comment:
Can we get away with not raising this exception to the user and only having the single `ValidationException` (or alternatively only having this exception not the other). I understand this is a subclass of that, but I can't imagine when a user would only want to catch this.

Or maybe `ValidationErrors` should be kept and the `__init__` could take in `*errors` where an `error` is some `namedtuple` or simple `ValidationError` class that has the `datum`, `schema`, and `field` property with a `__str__` that builds the message and the `ValidationErrors` exception can build the overall message based on the concatenation of the errors. Something like the following:

```python
class ValidationErrors(Exception):
    def __init__(self, *errors):
        super(Exception, self).__init__(', '.join(error for error in errors))
        self.errors = errors

class ValidationError(object):
    def __init__(self, datum, schema, field):
        self.datum = datum
        self.schema = schema
        self.field = field

    def __str__(self):
        return '{field} is {datum} of type {given_type} expected {schema}'. \
            format(datum=self.datum,
                   given_type=type(self.datum),
                   schema=self.schema,
                   field=self.field)
```

Alternatively, `ValidationErrors` could have some classmethod helpers like so:

```python
class ValidationErrors(Exception):
    
    @classmethod
    def from_single(cls, datum, schema, field):
        return ValidationErrors('{field} is {datum} of type {given_type} expected {schema}'. \
            format(datum=datum,
                   given_type=type(datum),
                   schema=schema,
                   field=field)
        )

    @classmethod
    def from_many(cls, messages):
        return ValidationErrors(', '.join(message for message in messages))
```

What are your thoughts?

Reply:
I think I like the first, it might be possible to make things lighter weight too with slots or namedtuple class for the ValidationError object.